File: lab9/weapon_package/water_pistol.py
"""
Модуль, що містить похідний клас WaterPistol.
"""
from pistol import Pistol

class WaterPistol(Pistol):
    """Клас, що описує водяний пістолет, який успадковує Pistol."""

    def __init__(self, model, capacity, color):
        """
        Конструктор похідного класу.
        :param color: Колір пластику
        """
        # Виклик конструктора базового класу
        super().__init__(model, capacity)
        self.color = color

    def shoot(self):
        """Перевизначений метод пострілу для води."""
        if self.ammo_count > 0:
            self.ammo_count -= 1
            print(f"{self.model} ({self.color}): Пс-с-ст! Виприснуто воду. Залишок: {self.ammo_count} мл")
        else:
            print(f"{self.model}: Вода закінчилася! Потрібна дозаправка.")

    def fill_full(self):
        """Специфічний метод для водяного пістолета."""
        self.ammo_count = self.capacity
        print(f"{self.model}: Повністю заправлений водою ({self.capacity} мл).")

File: LAB_03/src/ki305/Sadova/lab3/WaterPistol.java
package ki305.Sadova.lab3;

import java.io.FileNotFoundException;

/**
 * Клас <code>WaterPistol</code> розширює <code>Pistol</code> та реалізує <code>Refillable</code>.
 * Моделює роботу водяного пістолета.
 */
public class WaterPistol extends Pistol implements Refillable {

    private String waterType; // Наприклад, "Прісна" або "Солона"

    /**
     * Конструктор водяного пістолета.
     * @param capacity Ємність резервуара
     * @param waterType Тип води
     * @throws FileNotFoundException
     */
    public WaterPistol(int capacity, String waterType) throws FileNotFoundException {
        super();
        this.magazine = new Magazine(capacity); // Резервуар для води
        this.waterType = waterType;
        logger.log("WaterPistol created. Water type: " + waterType);
    }

    /**
     * Реалізація пострілу (випорскування води).
     */
    @Override
    public boolean shoot() {
        if (!trigger.isCocked()) {
            logger.log("WATER SHOOT FAILED: Pump it first (cock)!");
            return false;
        }
        if (magazine.getRounds() <= 0) {
            logger.log("WATER SHOOT FAILED: Out of water.");
            return false;
        }

        magazine.unload(1); // Витрачаємо "порцію" води
        trigger.reset();
        barrel.registerShot();
        logger.log("SQUIRT! Water shot successful. Remaining water: " + magazine.getRounds());
        return true;
    }

    @Override
    public void refill(int volume) {
        int added = magazine.load(volume);
        logger.log("Refilled " + added + " units of " + waterType + " water.");
    }

    @Override
    public void empty() {
        magazine.unload(magazine.getRounds());
        logger.log("Reservoir emptied.");
    }
    

    /**
     * Повертає тип води, що використовується.
     * @return waterType
     */
    public String getWaterType() {
        return waterType;
    }
}